### DISCUSSION TITLE
Using multiple observation files in Point-Stat

### DISCUSSION THREAD

**Discussion by DeannaSpindler-NOAA** [dtcenter/METplus#1747] [File I/O] [answered]
Hello,
I'm trying to use the NDBC buoy data to validate the GFS-Wave model using Point-Stat (METplus 10.1.1 and MET 4.1.1). The observation files are available either as NetCDF or as ASCII files that contain many observation times, but are one file per buoy and I need to use hundreds of buoys for a given time window.  I tried to subset and concatenate the NetCDF files using ncks and ncrcat, but I was not able to produce the single file.  I tried using ascii2nc on the ASCII files, but it could not determine the file format.  Before I attempt to use Python embedding with ascii2nc, is there some way to pass a list of buoy ID's to Point-Stat so it uses the individual buoy NetCDF files and calculates the stats as if I had passed in one file with all the buoys? Can you suggest any other way to proceed?
I have some sample buoy files on Hera: /scratch2/NCEPDEV/ocean/Deanna.Spindler/noscrub/NDBC
Many thanks,
Deanna

**Comment by JohnHalleyGotway** [comment]
@DeannaSpindler-NOAA, I see you're looking for some advice for how to handle the buoy point observations.

You've already found that these are not in a format currently supported by ascii2nc. With that in mind, using python embedding is definitely how I'd recommend proceeding. If the buoy point observations follow a well-establish, well-documented format, we certainly could consider enhancing ascii2nc to support them in the future. I'd recommend you bring this up for discussion at the next METplus NOAA User telecon.

But in the meantime, we should be able to get rolling with python embedding.

It would seem intuitive to me to write a `read_buoy.py` script that takes an arbitrary number input ascii files as input. It'd just read/reformat the data for each one and pass the result as one big block in memory to the MET tool you're calling.

As long as we encode them all using the same "message_type" then Point-Stat will group them all together in the computation of statistics. I'm headed out for the day, but let me know how much help you'd like getting started with this script.

Thanks.

John

**Comment by j-opatz** [comment]
Hi @DeannaSpindler-NOAA,

@JohnHalleyGotway mentioned in a recent meeting that I should share some "in progress" work that may be of interest to you.  

I'm working on a project for OPC to verify Wave Watch 3 output against observational datasets, expanding both satellite and buoy data. I'll link the Github issue here: https://github.com/dtcenter/METplus/issues/1482

So far, the use case is using the method John HG outlined in his answer: a Python script is retrieving ASCII buoy data for 200+ buoys and converting it into a single netCDF. That python script was provided to us by an intern that was working on a separate project that did not utilize METplus for verification, so the conglomerate netCDF is currently unreadable by METplus. I'm working to change that, with the end goal being a Python script that retrieves data from the NDBC, and converts the ASCII data into a memory-held data format that can be passed to Point-Stat for processing.

John HG and I both agree that while this should provide some direction/outright solution to the buoy issue you've described, it'd be best practice to create a protocol in ASCII2NC to ingest buoy data natively, especially when buoy data is QC'd and has a standard format.

**Reply by DeannaSpindler-NOAA** [reply]
Hello @j-opatz and @JohnHalleyGotway,

I am definitely interested in what you develop for OPC, both for satellite and buoy verification, as it is quite similar to what I need for EMC.  Do you suggest that I begin using the Python script that was provided to you by the intern, or should I wait for you to create a protocol in ASCII2NC to ingest the NDBC buoy data? If so, do you have an estimate for when you expect to have it ready?

Thanks,
Deanna

**Comment by j-opatz** [comment]
Hi Deanna,

That's great to hear that our goals align so closely with each other's! 

I would suggest that the best way to avoid duplicate work would be to wait until I have the Python embedding script worked out. That script will enable users (including yourself) to download buoy/satellite data and pass through memory whatever buoy/satellite data sets are required to METplus for evaluation.

While it's currently still a work in progress, I'll place the finalized script and whatever else is available at the time into a working Github branch that you can check out and use however you'd like. That part should be complete within the next two weeks.

The ascii2nc enhancement is more of a stretch goal, and while John HG is more able to comment on how difficult that might be to implement in MET, I know that the budget and time to do so is very scarce and may have to wait until the next major release (i.e. MET v 11.1 or more). The work around (i.e. Python embedding) will be done much quicker than that.

**Reply by j-opatz** [reply]
Deanna,

I wanted to check in one last time to let you know this work is still ongoing. Progress has been made on ingesting buoy data via Python embedding, but other project deadlines and trying to get all of the necessary information out of the netCDFs and into a memory-held dataframe that MET recognizes has been challenging. The latest decision was to keep time interpolation in the Python script active, so the user can interpolate from the raw time intervals to user set intervals, hopefully allowing more data comparisons. As mentioned, I'll get the finalized script into a GitHub branch that you can check out once it's complete.

**Comment by DeannaSpindler-NOAA** [comment, marked as answer]
Thanks for the update, John.  I completely understand about other project
deadlines!
Best,
Deanna

--
Deanna Spindler, PhD
Physical Scientist III
Lynker at NOAA/NWS/NCEP/Environmental Modeling Center
NOAA Center for Weather and Climate Prediction

**Reply by j-opatz** [reply]
Hi Deanna,

I figured I'd provide a final update in the hopes that we can close out this Discussion.

As you're aware (by your help on the issue), [#2276](https://github.com/dtcenter/MET/issues/2276) is an issue that will let a user read in buoy data natively by the ASCII2NC tool. This is a great capability being added that I hope solves the immediate issue you've been having with reading buoy data.

As I noted in my METplus use case issue [here](https://github.com/dtcenter/METplus/issues/1482#issuecomment-1278187780), the biggest snag currently blocking the Python script is a limit on how many stations/buoys can be passed to MET. That has its own issue for investigation. However, this new ability to read in buoy data has meant a pivot in the use case creation: originally, I was going to read in (via Python embedding) buoy data in netCDF format that was past the 45 day window and therefore had undergone proper QC. That would be passed to PointStat for analysis of the Wave Watch 3 model for wave heights. Now, I'll break the process into two steps: the first is a call to ASCII2NC that uses the new native method of reading in buoy data, and the second is passing the output netCDF to PointStat for analysis.

This unfortunately means that the use case will use unQC'd data, and that the Python script will most likely not see repository release until the next major METplus development cycle, next year. 

If you'd like, I can send over what currently exists of the Python script for your use: it does work with Point-Stat on QCd buoy data in netCDF format, as long as the total number of buoys being used is equal to or less than 145 entries. I understand if that limitation is too great, and sincerely hope to have a solution in the near future.

**Comment by j-opatz** [comment]
Hi Deanna,

Another ping to see if you had any additional comments/questions for the work that's been done to ingest buoys. If you're satisfied, we can close out this discussion.

**Reply by j-opatz** [reply]
Given the length of time since the last comment from the original poster, the Discussion is being closed for inactivity. If new questions arise, please open a new Discussion and we'll respond there.
